include/verilated_random.cpp

```diff
@@ -0,0 +1,212 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+//
+// Code available from: https://verilator.org
+//
+// Copyright 2024 by Antmicro Ltd.  This program is free software; you can
+// redistribute it and/or modify it under the terms of either the GNU Lesser
+// General Public License Version 3 or the Perl Artistic License Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//=========================================================================
+///
+/// \file
+/// \brief Verilated randomization implementation code
+///
+/// This file must be compiled and linked against all Verilated objects
+/// that use randomization features.
+///
+/// See the internals documentation docs/internals.rst for details.
+///
+//=========================================================================
+
+#include "verilated_random.h"
+
+#include <fstream>
+#include <sstream>
+
+#include <sys/wait.h>
+
+#define HASH_LEN 1
+
+static const char* solvers[] = {
+    "z3 ",
+    "cvc5 --lang=smt2 --incremental ",
+    "cvc4 --lang=smt2 --incremental ",
+};
+
+static std::string get_solver() {
+    static std::string solver;
+    static bool found = false;
+    if (found) return solver;
+
+    for (const char* cmd : solvers) {
+        std::string comm = cmd + std::string{"/dev/null 2>/dev/null"};
+        int ec = system(comm.c_str());
+        if (!ec) {
+            solver = cmd;
+            found = true;
+            return solver;
+        }
+    }
+    std::stringstream msg;
+    msg << "No SAT solvers found, please install one of them.  Tried:\n";
+    for (const char* cmd : solvers) msg << "$ " << cmd << '\n';
+
+    VL_WARN_MT("", 0, "randomize", msg.str().c_str());
+
+    found = true;
+    return solver;
+}
+
+//======================================================================
+// VlRandomizer:: Methods
+
+void VlRandomVar::emit(std::ostream& s) const { s << m_name; }
+void VlRandomConst::emit(std::ostream& s) const {
```

Comment:
The constants and variables are just bit vectors, but comparisons, multiplication and such use separate operations (`bvsgt` vs `bvugt`, `bvsmul` vs `bvmul` etc). I added an assertion to the constructor just in case.